api/src/main/java/org/eclipse/microprofile/config/spi/ConfigProviderResolver.java

```diff
@@ -91,9 +91,23 @@ protected ConfigProviderResolver() {
      *
      * If the given Config is associated within an Application then it will be unregistered.
      * @param config the config to be released
+     *
+     * @deprecated This method is not adequately specified and can lead to problems when a configuration is registered
```

Comment:
No this statement is based on a fundamentally wrong assumption. A Config must at maximum be registered with a single Application (-> 1 ClassLoader)

Reply:
Here's the problem: I can't find anything in the specification that says this.

When you're dealing in specifications, everything in the specification is normative, which is to say "it's the truth".  Everything *not* in the specification is an opinion, even if it has historical significance.  It is not correct to say that the assumption is wrong, because until this proposed change, there's nothing in the specification to indicate whether it is wrong or right (hence the change).

So in order to resolve the issue, we have to update the specification in *some* way to make it consistent.  It may be your opinion that certain APIs were design with certain intent, but while that information is useful to feed into the decision process, it is secondary to other concerns, such as compatibility.

When I talk about "compatibility" I'm not just talking about it in the traditional sense of source or binary or even behavioral compatibility.  I'm talking about whether there is value in essentially cutting off or disabling existing uses and use cases that were established due to the interpretation of the specification *as it is written*, rather than *as it was originally intended to be*.  So while the correct remedy *might* be to "fix" the specification to meet the original intent, it also might *not* be the correct remedy given the ways in which the specification has already been adopted for usage.

So what I'm trying to say is that if, whether by intent or by error, something is left out of the specification, then that omission is now normative or "correct" by definition, and once it's out in the world one must assume that this normative behavior now cannot just be "fixed up" after the fact without consequences.  And in viewing those consequences one must consider existing users and use cases, and whether these can be justifiably invalidated (I believe the answer to this question is nearly always going to be "no" because use cases generally do not *become* invalid; they either are or are not, and sometimes we might make a mistake in that judgement).